robot_smach_states/src/robot_smach_states/navigation/guidance.py

```diff
@@ -30,30 +34,183 @@ def _detect_operator_behind_robot(robot, distance=1.0, radius=0.5):
     success, found_people_ids = robot.ed.detect_people(*image_data)
     found_people = [robot.ed.get_entity(id_) for id_ in found_people_ids]
 
+    rospy.loginfo("Found {} people: {}".format(len(found_people), found_people))
+    found_people = [p for p in found_people if p]
+    rospy.loginfo("{} people remaining after None-check".format(len(found_people)))
+
     # Assume the operator is around 1.0 m behind the robot
     base_pose = robot.base.get_location()
     expected_person_pos = base_pose.frame * kdl.Vector(-distance, 0.0, 0.0)
 
+    rospy.loginfo("\n\tBase pos, expected person pos: {}, {}".format(base_pose.frame.p, expected_person_pos))
     for person in found_people:
+        rospy.loginfo("\n\tPerson at {}, distance: {} (radius = {})".format(
+            person.pose.frame.p,
+            (person.pose.frame.p - expected_person_pos).Norm(),
+            radius
+        ))
         if (person.pose.frame.p - expected_person_pos).Norm() < radius:
             return True
     return False
 
 
+class TourGuide(object):
+    def __init__(self, robot, x_threshold=0.75, y_threshold=1.5):
+
+        self._robot = robot
+
+        # Properties
+        self._x_threshold = x_threshold
+        self._y_threshold = y_threshold
+
+        # State
+        self._furniture_entities = []
+        self._room_entities = []
+        self._furniture_entities_room = {}  # map room entities to furniture entities
+        self._passed_room_ids = []  # Will contain the ids of the rooms that are passed
+        self._passed_furniture_ids = []  # Will contain the ids of the furniture that is passed
+
+        self.initialize()
+
+    def describe_near_objects(self):
+        # type: () -> string
+        """
+        Describes near objects based on the robots current position
+        :return: A description of the robots surroundings
+        :rtype: string
+        """
+        position = self._robot.base.get_location()
+
+        try:
+            room = self.get_room(position.frame.p)
+        except RuntimeError:
+            rospy.logwarn("position ({}) not in any room".format(position.frame.p))
+            return ""
+
+        if room.id not in self._passed_room_ids:
+            self._passed_room_ids.append(room.id)
+            rospy.logdebug("describe entering room: {}.\t passed rooms is now {}".format(room.id, self._passed_room_ids))
+            return "We now enter the {}".format(room.id)
+
+        # not entering a new room, checking furniture
+        furniture_objects = self._furniture_entities_room[room]  # Furniture objects in our current room
+        for entity in furniture_objects:  # type: Entity
+            # Check if in passed ids
+            if entity.id in self._passed_furniture_ids:
+                continue
```

Comment:
What is the use of this if? It passes whether this condition is satisfied or not.

Reply:
this continue breaks out of the for loop. Thereby preventing the robot from describing a piece of furniture multiple times